File: backend/app/models/product.py
"""SQLAlchemy ORM model for the products table."""

import uuid
from datetime import datetime, timezone
from typing import Any

from sqlalchemy import String, Boolean, DateTime, ForeignKey
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class Product(Base):
    __tablename__ = "products"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String, nullable=False)
    brand: Mapped[str | None] = mapped_column(String, nullable=True)
    ingredients: Mapped[list[Any]] = mapped_column(JSONB, nullable=False, default=list)
    image_url: Mapped[str | None] = mapped_column(String, nullable=True)
    is_catalog: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=_utcnow
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, default=_utcnow, onupdate=_utcnow
    )

File: mobile/src/db/models/Product.ts
import {Model} from '@nozbe/watermelondb';
import {field, date, text} from '@nozbe/watermelondb/decorators';

export class Product extends Model {
  static table = 'products';

  @text('name') name!: string;
  @text('brand') brand!: string;
  @text('ingredients') ingredients!: string; // JSON string
  @text('image_url') imageUrl!: string;
  @field('is_catalog') isCatalog!: boolean;
  @date('created_at') createdAt!: Date;
  @date('updated_at') updatedAt!: Date;

  // Sync metadata
  @text('_status') syncStatus!: string;
  @text('_changed') syncChanged!: string;
  @text('server_id') serverId!: string;

  get ingredientList(): string[] {
    try {
      return JSON.parse(this.ingredients) as string[];
    } catch {
      return [];
    }
  }
}
